enhancements/update/targeted-update-edge-blocking.md

```diff
@@ -0,0 +1,400 @@
+There is a risk, with the dynamic, per-cluster graph, that targeted edge blocking removes all outgoing update recommendations for some clusters on supported releases.
+This risk can be mitigated in at least two ways:
+
+* For the fraction of customer clusters that do not [opt-out of submitting Insights/Telemetry][uploaded-telemetry-opt-out], we can monitor [the existing `cluster_version_available_updates`][uploaded-telemetry-cluster_version_available_updates] to check for clusters running older versions which are still reporting no available, recommended updates.
+
+* We can process the graph with tooling that removes all `conditionalEdges` and look for any supported versions without exit paths.
+
+## Design Details
+
+### Open Questions
+
+#### Metadata to include when blocking a conditional request
+
+A URI seems like the floor here.
+But do we want other information as well?
+For example, a reason and message so we can say:
+
+```Text
+Reason: InvalidRouteHostSupport
+Message: Updating to 4.7.12 is supported, but not recommended, because your cluster contains routes with invalid spec.host.  For more details, see https://bugzilla.redhat.com/show_bug.cgi?id=1957584#c19
+```
+
+The benefit would be that folks would have some context about what they'd see if the clicked through to the detail URI.
+The downside would be that we'd need to boil the issue down to a slug and sentence or two, and that might lead to some tension about how much detail to include.
+
+#### Non-PromQL filters
+
+Teaching the CVO to query the local Thanos/Prometheus shouldn't be too bad (openshift/origin already does this).
+But to simplify things out of the gate, we may want to use `platform` or something that has less granularity, but simpler execution (the CVO would pull the Infrastructure resource and compare with the configured `platform` to decide if the edge was recommended).
+Worth doing this in 1.1.0, and extending to PromQL in a backwards-compatible 1.2.0?
+
+#### Reporting recommendation freshness
+
+Currently [`availableUpdates`][api-availableUpdates] does not have a way to declare the freshness of its contents (e.g. "based on data retrieved from the upstream update service at `$TIMESTAMP`").
+We do set the `RetrievedUpdates` condition and eventually alert if there are problems retrieving updates, and the expectation is that if we aren't complaining about being stale, we're fresh enough.
+We could take the same approach with `conditionalUpdates`, but now that we also have "based on evaluations of the in-cluster state at `$TIMESTAMP`" in the mix, we may want to proactively declare the time.
+On the other hand, continually bumping something that's similar to the node's `lastHeartbeatTime` is a bunch of busywork for both the cluster-version operator and the API-server.
+Is the added transparency worth it?
```

Comment:
We could worry only about freshness of the graph from the update service, silently re-evaluate on some reasonable schedule, but then also re-evaluate on a spec change and if there's a diff between the two evaluations on the relevant edge throw an error about resource being out of date? I've often thought that we should force a refresh on spec change even with current implementation.

Reply:
refresh on spec change, withholding the `observedGeneration` until the refresh is complete... Hmm... if we refresh everything from the update-service poll through to the condition evaluations for any conditional edges, that could take a while.  I think the current behavior is based on the expectation that both the update service metadata and the relevant in-cluster configurations are slowly-changing, on the order of hours, vs. our poll periods, on the order of minutes.  We can see how long the full refresh takes, and then decide if we want to gate `spec` change acceptance on that whole cycle.